synapse/lib/trigger.py

```diff
@@ -182,6 +203,104 @@ async def runTagDel(self, node, tag, view=None):
                 for _, trig in globs.get(tag):
                     await trig.execute(node, vars=vars, view=view)
 
+    async def runEdgeAdd(self, n1, verb, n2, view=None):
+        n1form = n1.form.name if n1 else None
+        n2form = n2.form.name if n2 else None
+        vars = {'auto': {'opts': {'verb': verb, 'n2': n2}}}
+        with self._recursion_check():
+            cachekey = (n1form, verb, n2form)
+            cached = self.edgeaddcache.get(cachekey)
+            if cached is None:
+                cached = []
+                for trig in self.edgeadd.get((None, verb, None), ()):
+                    cached.append(trig)
+
+                globs = self.edgeaddglobs.get((None, None))
+                if globs:
+                    for _, trig in globs.get(verb):
+                        cached.append(trig)
+
+                if n1:
+                    for trig in self.edgeadd.get((n1form, verb, None), ()):
+                        cached.append(trig)
+
+                    globs = self.edgeaddglobs.get((n1form, None))
+                    if globs:
+                        for _, trig in globs.get(verb):
+                            cached.append(trig)
+
+                if n2:
+                    for trig in self.edgeadd.get((None, verb, n2form), ()):
+                        cached.append(trig)
+
+                    globs = self.edgeaddglobs.get((None, n2form))
+                    if globs:
+                        for _, trig in globs.get(verb):
+                            cached.append(trig)
+
+                if n1 and n2:
+                    for trig in self.edgeadd.get((n1form, verb, n2form), ()):
+                        cached.append(trig)
+
+                    globs = self.edgeaddglobs.get((n1form, n2form))
+                    if globs:
+                        for _, trig in globs.get(verb):
+                            cached.append(trig)
+
+                self.edgeaddcache[cachekey] = cached
+
+            for trig in cached:
+                await trig.execute(n1, vars=vars, view=view)
+
+    async def runEdgeDel(self, n1, verb, n2, view=None):
+        n1form = n1.form.name if n1 else None
+        n2form = n2.form.name if n2 else None
+        vars = {'auto': {'opts': {'verb': verb, 'n2': n2}}}
```

Comment:
If `n2` is a heavy Node object then we cannot run this as an async trigger. If we converted n2 into a node iden then it could be run as async trigger and it would then the be responsibility of the trigger author to lift the node as needed.